File: Graph/3503/3503.py
source, target, n = map(int, input().split())
nums = list(map(int, input().split()))

LIMIT = 100000

def BFS(source, target):
    if source == target:
        return 0

    visited = [False] * (LIMIT + 1)
    visited[source] = True

    queue = [(source, 0)]
    head = 0

    while head < len(queue):
        value, operations = queue[head]
        head += 1

        if value == target:
            return operations

        for num in nums:

            # Soma
            x = value + num
            if 1 <= x <= LIMIT and not visited[x]:
                if x == target:
                    return operations + 1
                visited[x] = True
                queue.append((x, operations + 1))

            # Subtração
            x = value - num
            if 1 <= x <= LIMIT and not visited[x]:
                if x == target:
                    return operations + 1
                visited[x] = True
                queue.append((x, operations + 1))

            # Multiplicação
            x = value * num
            if 1 <= x <= LIMIT and not visited[x]:
                if x == target:
                    return operations + 1
                visited[x] = True
                queue.append((x, operations + 1))

            # Divisão
            if num != 0 and value % num == 0:
                x = value // num
                if 1 <= x <= LIMIT and not visited[x]:
                    if x == target:
                        return operations + 1
                    visited[x] = True
                    queue.append((x, operations + 1))

    return -1


print(BFS(source, target))

File: Graph/3503/3503.php
<?php

const LIMIT = 100000;

function BFS($source, $target, $nums)
{
    $visited = array_fill(0, LIMIT + 1, false);

    $queue = [];
    $head = 0;

    $queue[] = [$source, 0];
    $visited[$source] = true;

    while ($head < count($queue)) {

        [$curr, $step] = $queue[$head++];

        if ($curr == $target)
            return $step;

        foreach ($nums as $a) {

            // Soma
            $next = $curr + $a;
            if ($next <= LIMIT && !$visited[$next]) {
                $visited[$next] = true;
                $queue[] = [$next, $step + 1];
            }

            // Subtração
            $next = $curr - $a;
            if ($next >= 1 && !$visited[$next]) {
                $visited[$next] = true;
                $queue[] = [$next, $step + 1];
            }

            // Multiplicação
            $next = $curr * $a;
            if ($next <= LIMIT && !$visited[$next]) {
                $visited[$next] = true;
                $queue[] = [$next, $step + 1];
            }

            // Divisão
            if ($curr % $a == 0) {
                $next = intdiv($curr, $a);

                if (!$visited[$next]) {
                    $visited[$next] = true;
                    $queue[] = [$next, $step + 1];
                }
            }
        }
    }

    return -1;
}

[$X, $Y, $N] = array_map('intval', explode(' ', trim(fgets(STDIN))));
$nums = array_map('intval', explode(' ', trim(fgets(STDIN))));

echo BFS($X, $Y, $nums) . PHP_EOL;